### ISSUE TITLE
enlargethispage + twocolumn + footnote + float = wrong column

### ISSUE THREAD

**Issue by urrameu** [latex3/latex2e#399]
## Brief outline of the bug

In a two column text, `\enlargethispage` in the first column may very rarely break the column at the right place but actually enlarge/shrink the second column instead of the first, with very bad results. Apparently, this happens when there are both a float and a footnote involved. Sizes are crucial (which is why I did not use `lipsum` for the example text): In the MWE below, changing the value of -460pt or adding/removing lines solves the problem, but I experimented with different amounts of text and was always able to find a value for `\enlargethispage` that would bring the issue back.

I am pretty sure this hardly affects anyone, but I would like to know if it is possible to at least work around the issue: I am writing code that automatically calculates and inserts some negative `\enlargethispage`, which might cause this problem.

Thanks!

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}
\documentclass[twocolumn]{article}
\begin{document}

\enlargethispage{-460pt}

\section{Some section}

This is the beginning of a meaningless paragraph. Some meaningless text.
Some meaningless text. Some meaningless text.\footnote{This is the
beginning of a meaningless footnote. Some meaningless footnote text.
Some meaningless footnote text. Some meaningless footnote text.}

\begin{figure}[htbp]
  \fbox{\parbox[c][4\baselineskip][c]{.9\columnwidth}
    {\hfill This is a figure\hfill\null}}
\end{figure}

This is the beginning of a meaningless paragraph. Some meaningless text.

\end{document}
```

## Log file (required) and possibly PDF file  

[enlargethispage.log](https://github.com/latex3/latex2e/files/5319630/enlargethispage.log)
[enlargethispage.pdf](https://github.com/latex3/latex2e/files/5319631/enlargethispage.pdf)

**Comment by FrankMittelbach** [comment]
woah. If you look in the document latex code (around `\@reinserts`) you will find a comment dating back to 1993 that we think that this code might cause issues in some cases (but that we don't quite know what these cases might be). Looks like after 27 years you finally found an MWE that proves our prediction.

I think I understand what goes wrong in your case, but I'm not yet sure if it can be corrected without possibly generating harm to many more documents that worked flawlessly during the last 3 decades -- we'll see.

**Comment by car222222** [comment]
I thought that this might be coming my way:-)!

My memory is hazy on this but I do not recall any thought about ensuring that it works for columns.  But I am not saying that this usage is the reason for the (possible:-) bug.

**Comment by FrankMittelbach** [comment]
Nothing to do with columns as far as I can see, I'm sure you can simplify the MWE to a single column case. In the nutshell, what happens is this: \enlargethispage creates an insert. and so does \footnote. But the footnote insert is too big to fit and will split. However, LaTeX doesn't immediately does this because the footnote is in the last para before the figure, so the next thing that happens is that the figure is seen and handled. Handling means it gets deferred, but when returning from the OR the placement of the 2 inserts are both moved below the current text. As a result TeX now tries to split (not the footnote insert but the enlargepage insert (which it can't) and therefore is forced to settle on an earlier break than it should.

**Comment by car222222** [comment]
Aha yes! footnotes are likely cause problems because both they and this command (surprise!) use inserts.

**Comment by FrankMittelbach** [comment]
A safe solution (I think) can only be to use `\holdinginserts=1` while doing the float handling to avoid any change in insert positions when putting stuff back. That wasn't available in the early days and not well understood in 1993 and I'm not sure it can be put into the current model without causing issue elsewhere. This is why I said I'm not sure it can be fixed.

**Comment by FrankMittelbach** [comment]
Or maybe not. Try the following if that works for you
```
\makeatletter
\gdef \@reinserts{%
  \ifvbox\@kludgeins\insert\@kludgeins
                      {\unvbox\@kludgeins}\fi
  \ifvoid\footins\else\insert\footins{\unvbox\footins}\fi
}
\makeatother
```
I haven't checked if this is causing harm elsewhere but I guess not (though my guess have been wrong in the past).

**Comment by urrameu** [comment]
> Looks like after 27 years you finally found an MWE that proves our prediction.

Hmmm, better late than never? :-p I actually wrote a script that generates random LaTeX documents searching for problems with my code and this appeared only once in a couple thousand iterations.

> Nothing to do with columns as far as I can see, I'm sure you can simplify the MWE to a single column case.

I tried quite a bit to adapt the example to a single column without success before reporting.

> Try the following if that works for you

It does work, at least for this example, thanks!

I will (1) try a little harder to find a MWE without columns and (2) run my test script with this modification applied. I'll report my results here in about a week.

**Comment by FrankMittelbach** [comment]
my suggested fix isn't correct, this goes deeper unfortunately and has to do with the way TeX splits insertions (and looses information along the way. So this is a) more complicated and b) most likely confined to a combination of \enlargethispage + a footnote that doesn't fit onto the page + a float that is on that page. So this needs further analysis (nevereless additional MWes would be welcome)

**Comment by car222222** [comment]
"nevereless" ??  Frank really meant "therefore", despite trying to write "nevertheless" (took me three times to type it right!) :-)

**Comment by FrankMittelbach** [comment]
well typing is difficult ... yes I mean additional examples that go wrong will help in the anlysis

**Comment by urrameu** [comment]
Some more information:

1. Obviously, you were correct: the problem appears in single-column documents too. A variation of the previous MWE in single-column:

```latex
\RequirePackage{latexbug}
\documentclass{article}
\begin{document}

\enlargethispage{-460pt}

\section{Some section}

This is the beginning of a meaningless paragraph. Some meaningless text.
Some meaningless text.  Some meaningless text.  Some meaningless text.
Some meaningless text. Some meaningless text.\footnote{This is the
beginning of a meaningless footnote. Some meaningless footnote text.
beginning of a meaningless footnote. Some meaningless footnote text.
Some meaningless footnote text. Some meaningless footnote text.}

\begin{figure}[htbp]
  \fbox{\parbox[c][4\baselineskip][c]{.9\columnwidth}
    {\hfill This is a figure\hfill\null}}
\end{figure}

This is the beginning of a meaningless paragraph. Some meaningless text.
Some meaningless text. Some meaningless text. Some meaningless text.

\end{document}

```

2. I created 6,000 random documents using `\enlargethispage` and got about one every 500 that showcased this bug (my test files use two columns to make automatic detection easier). In all but one case, this bug happened when:
 a. There was a footnote in the last line of the last paragraph of the first column that needed to be split
 b. There was a float defined after this footnote (its size is not relevant)

3. When the bug is triggered, if there is another footnote in the first column, it gets moved to the second column (this example is particularly interesting because it shows the bug may appear with `\enlargethispage{-3pt}`, check the commented line):

```latex
\documentclass[twocolumn]{article}
\flushbottom
\begin{document}

\section{Some section}
This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\footnote{This is the beginning of a meaningless footnote.}\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\par

\enlargethispage{-200pt}
%\enlargethispage{-3pt}\vspace{190pt}

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\par

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\footnote{This is the beginning of a meaningless footnote.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.}\par

\begin{figure}[htbp]
  \centering
  \fbox{%
    \parbox[c][12\baselineskip][c]{.9\columnwidth}
    {\hfill This is a figure\hfill\null}%
  }
\end{figure}

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\par

\end{document}
```

4. In all such cases, the fix proposed in [the comment above (num. 6)](#issuecomment-703075465) works, **but** in the example below the footnote appears in a different column from the footnote mark:

```latex
\documentclass[twocolumn]{article}
\makeatletter
\gdef \@reinserts{%
  \ifvbox\@kludgeins\insert\@kludgeins
                      {\unvbox\@kludgeins}\fi
  \ifvoid\footins\else\insert\footins{\unvbox\footins}\fi
}
\makeatother
\flushbottom
\begin{document}

\enlargethispage{-80pt}

\section{Some section}
This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\par

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\footnote{This is the beginning of a meaningless footnote.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.}\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\par

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\footnote{This is the beginning of a meaningless footnote.\\
Some meaningless footnote text.}\par

\begin{table}[htb]
  \centering
  \fbox{%
    \parbox[c][5.7\baselineskip][c]{.9\columnwidth}
    {\hfill This is a table\hfill\null}%
  }
\end{table}
\end{document}
```

Finally, I found only one failing case that did not involve footnotes (it involves different float types instead). I tried to simplify the example as much as possible, but it is still quite convoluted:

```latex
\documentclass[twocolumn]{article}
\flushbottom
\usepackage{lipsum}
\begin{document}

\enlargethispage{-59pt}

\lipsum[1-2]

% This glue is essential. It can appear anywhere in this column
% And should be "near" the size of \enlargethispage
\vspace{0pt plus 50pt minus 4pt}

\begin{figure}[htbp]
  \centering
  \fbox{%
    \parbox[c][12\baselineskip][c]{.9\columnwidth}
    {\hfill A \textbf{figure} that does not fit ``here''\hfill\null}%
  }
\end{figure}

\begin{table}[htbp]
  \centering
  \fbox{%
    \parbox[c][10\baselineskip][c]{.9\columnwidth}
    {\centering A \textbf{table} that \textbf{does} fit ``here''\par}%
  }
\end{table}

\begin{figure}[htbp]
  \centering
  \fbox{%
    \parbox[c][2\baselineskip][c]{.9\columnwidth}
    {\hfill Some float\hfill\null}%
  }
\end{figure}

\begin{figure}[htbp]
  \centering
  \fbox{%
    \parbox[c][2\baselineskip][c]{.9\columnwidth}
    {\hfill Some other float\hfill\null}%
  }
\end{figure}

This is the beginning of a meaningless paragraph. Some meaningless text.\par

\end{document}
```

For this version of the problem to happen, there has to be some glue. I added some manually but, depending on the size of things, the default glue around floats is enough.

The proposed fix does not work with this final example, but it does help with the others; I dare say it is a nice addition, even if not a perfect solution.

**Comment by urrameu** [comment]
Sorry for spamming the bug report, but I have yet a little more information:

1. I found another example of the "different" problem (this really is *very* rare to come by!). This example involves one footnote and one float, but the footnote is **not** on the last line of the paragraph and there has to be some glue:

```latex
\documentclass[twocolumn]{article}
\flushbottom
\begin{document}

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\par

\enlargethispage{-104pt}\vspace{0pt minus 2pt}

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\footnote{This is the beginning of a meaningless footnote.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.\\
Some meaningless footnote text.}\\
Some meaningless text.\par

\begin{figure}[htbp]
  \centering
  \fbox{%
    \parbox[c][12\baselineskip][c]{.9\columnwidth}
    {\hfill This is a figure\hfill\null}%
  }
\end{figure}

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\par
\end{document}
```

2. I generated a few thousand documents with `\enlargethispage`, compiled each one with and without the proposed fix from comment 6 and compared the results. Besides the documents affected by this bug, some others compiled differently: with the proposed modification, LaTeX is sometimes more willing to break a footnote across columns. An example:

```latex
\documentclass[twocolumn]{article}
\flushbottom
%\makeatletter
%\gdef \@reinserts{%
%  \ifvbox\@kludgeins\insert\@kludgeins
%                      {\unvbox\@kludgeins}\fi
%  \ifvoid\footins\else\insert\footins{\unvbox\footins}\fi
%}
%\makeatother
\begin{document}

\enlargethispage{-249pt}

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\par

This is the beginning of a meaningless paragraph.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\footnote{This is the beginning of a meaningless footnote.\\
Some meaningless footnote text.}\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\\
Some meaningless text.\par

\begin{figure}[htb]
  \centering
  \fbox{%
    \parbox[c][5\baselineskip][c]{.9\columnwidth}
    {\hfill This is a figure\hfill\null}%
  }
\end{figure}

This is the beginning of a meaningless paragraph.\par
\end{document}
```

While this difference does not indicate a defect, it means incorporating this change could very rarely modify existing documents. Since I do not have to worry about that, this "fix" basically solves my particular problem, even if the actual issue is more complicated, as suggested by FrankMittelbach.

**Comment by FrankMittelbach** [comment]
> Sorry for spamming the bug report, but I have yet a little more information:

I wouldn't consider this spamming, it is quite useful information to figure out what is going to be the best cause of action. Basically it should that my original suggestion is not actually curing all cases  and that there is something else that is not quite right: my suspicion is that somehow the calculation is off by the depth output  box that is constructed when the figure is handled. For example, if you remove the descender in the last line of your second example, i.e.,
```
Some meaningless text.\\
Some meaninless text.\par         %<--- g droppped

\begin{figure}[htb]
```
I need to figure out why that is the case and whether that is a general error in the way the insertion boxes are handled in the special OR.

**Comment by car222222** [comment]
But do not forget that there are anyway known problems in LaTeX's handling of footnotes when combined with floats and marginals, even without \enlargethispage.

**Comment by urrameu** [comment]
I reported four different problem cases:

1. `\enlargethispage` affects the wrong column/page because of a split footnote in the last line of the last paragraph of the first page followed by a float (my original report);

2. `\enlargethispage` affects the wrong column/page in some obscure situation involving glues and the depth of the last line (the "g" you mentioned). I only found two examples for this case, one involving footnotes with floats and the other involving different float types;

3. In some cases, LaTeX may split footnotes differently when using the proposed fix;

4. A footnote may end up on the wrong column/page when using the proposed fix (I only found one example of this).

I believe there are two different bugs at play here; based on what you wrote, let's call the first "wrong insert split" and the second "wrong depth calculation". Case 1 above is caused by the "wrong insert split" bug and seems to be completely solved by your proposed fix. Cases 2 and 3 are caused by the "wrong deph calculation" bug, since removing the letter g makes them disappear. Removing the letter g also solves case 4, which makes me think it is a combination of both (that explains why I only found one example: you must have really bad luck to trigger two bugs at once).

Why do I think these are different bugs?

1. Cases 1 and 2 are triggered in different situations;

2. Case 1 is solved by the proposed fix while case 2 is not;

3. In case 1, small changes to the size of `\enlargethispage` and the number of lines etc. still generate a defective document, i.e., the problem involves somewhat "large" dimensions. In case 2, even minuscule changes make the problem disappear, suggesting very small dimensions are at play.

However, as I mentioned, for new documents the proposed fix almost eliminates any chance of problems.

**Comment by FrankMittelbach** [comment]
> Why do I think these are different bugs?

I'm not sure you are right here, I think it is the same underlying bug. The reason that changing the order of reinserting the inserts make a difference for some cases is due to the fact that footnote boxes can split (and so absorb an incorrect calculated "extra" through splitting) while the enlargethispage insert is a single block which can only move to the next column or page if it doesn't fit. In other words, I think the order should be irrelevant it the calculations are done correctly.
Just changing the order fixes some cases and alters some others (and makes some of them wrong, e.g., getting a split footnote when the callout of the footnote is in the middle of the page is rather odd and against traditional typesetting rules.

Anyway, this is no here or there and needs further analysis. Maybe changing the order is all we can do to get "more" cases right, but I think understanding where any why the calculations are going wrong is the real key to solving the issue.

**Comment by car222222** [comment]
Maybe it would be useful to write out the algorithm/processing that should be applied in each case, both those that are processing an insert and all the cases that call the LaTeX OR.

This might reveal problems that are not due to incorrect coding.  For example, cases where the abstractly optimal action is not in practice feasible.

**Comment by car222222** [comment]
Apologies if this question has already been answered: are there examples of each of the following?

Moving the 'enlargethispage' command gets the correct output.
Moving it does not help.

**Comment by urrameu** [comment]
> I'm not sure you are right here, I think it is the same underlying bug.

I trust your instinct on this :) .

I did find out something that I believe is very relevant: I tried to verify whether removing the descenders would make the problem disappear from the examples. This does not happen in the general case (but it does in some). What is interesting, though, is that apparently LaTeX **is** splitting the footnote in the right place, but then it typesets both slices together. Check this example out (it is the first example I posted, but without ascenders/descenders in the footnote to make my point more obvious):

```latex
\documentclass[twocolumn]{article}
\begin{document}
\enlargethispage{-445pt}
\section{Some section}

This is the beginning of a meaningless paragraph. Some meaningless text.
Some meaningless text. Some meaningless text.\footnote{xxxs xs xxe
xexxnnxnx ox a meanxnxxess xooxnoxe. some meanxnxxess xooxnoxe xexx.
xexxnnxnx ox a meanxnxxess xooxnoxe. some meanxnxxess xooxnoxe xexx.
xexxnnxnx ox a meanxnxxess xooxnoxe. some meanxnxxess xooxnoxe xexx.
some meanxnxxess xooxnoxe xexx. some meanxnxxess xooxnoxe xexx.}

\begin{figure}[htbp]
  \fbox{\parbox[c][4\baselineskip][c]{.9\columnwidth}
    {\hfill This is a figure\hfill\null}}
\end{figure}

This is the beginning of a meaningless paragraph. Some meaningless text.
\end{document}
```

Notice that the interline spacing in the footnote is wrong. This coincides with where the footnote is split when using the modification from comment 6. In the example where the fix causes the footnote to appear in the second column but the footnotemark to appear in the first, the same happens with and without the fix, suggesting that LaTeX tried to split the footnote in the same (and correct) place both times (and suggesting that, indeed, the "fix" is not really a fix). The same is true for the example that I called the "different" case with a footnote. What does all that mean? No idea :)

I also noticed that, in the only example that does not involve footnotes, changing the position of the gue on the page makes the problem go away, i.e., the glue needs to be at the bottom of the page. I imagined that eliminating the elasticity from `\skip\footins` would affect the other examples, but that did not happen (I even removed all glues with no effect). Again, no idea what does that mean.

> Moving the 'enlargethispage' command gets the correct output.
> Moving it does not help.

In all examples, I tried moving `\enlargethispage` around and that had no effect as long as it stays within the first column/page (if `\enlargethispage` appears after the footnote, it is already on the second column/page).

**Comment by u-fischer** [comment]
I'm not sure if it is related (as no `\enlargethispage` is involved). But from a tex.sx question I made this example where the footnote is pushed to the wrong side too:

~~~~
\documentclass[a4paper,11pt]{article}

\begin{document}

dag\\dag

\vspace*{28\baselineskip}

dag\\dag\\dag%
\footnote{I would like to have the footnote text on the same page as the footnote mark.}


\section{New Section}

\begin{table}
table\\table\\table\\table\\table
\end{table}

blub

\end{document}
~~~~

**Comment by FrankMittelbach** [comment]
@u-fischer no, this is not related to `\enlargethispage`. It is just a fact of the standard way the algorithm behaves: Whenever LaTeX encounters a float it calls on the output routine to settle its placement. After that the OR returns the page material back to the main vertical list but there will be one difference:
 - before the call to the OR  the footnotes are individually located next to the place they  are encountered
 - after the OR has been called they are all bunched together in the insertion box.
 - so LaTeX has to make a decision where to place them, eg first or last. currently it is last 

So in your example they are reinserted after the `\section`, but then LaTeX determines to break before the section, hence the result.

Doing it first has the opposite problem (and one that would happen more requently I guess) that you might pick up a footnote text where the mark then ends up on the  following page.

For this problem the correct approach is to use `\holdinginserts` when calling the OR but this is not easy to integrate into the algorithm as it is a major change.

**Comment by car222222** [comment]
Indeed, I used to have a large museum of ways to confuse the footnote handling.

We did do a lot of thinking about improving it when 2e came along but, as Frank implied, it seemed at that stage to be too much to take on.  

Marginpars can also interfere with all this and, for different reasons, appear in the wrong margin or get lost completely.

**Comment by FrankMittelbach** [comment]
Part of the solution could be as simple as this:
```
\makeatletter
\patchcmd\end@float{\@cons\@currlist\@currbox}
   {\@cons\@currlist\@currbox
    \global\holdinginserts\@ne}
    {}{}
    
\apptocmd\@specialoutput{\global\holdinginserts\z@}

\makeatother
```
which would solve the problem that @u-fischer mentioned. However, this is not enough because for the page enlargement you need the information about the insertion content, so you would need both a version with and a version without `\holdinginserts` or alternatively a totally different implementation for `\enlargethispage`, either makes it difficult to do safely.

@car222222 that marginpars can come out wrongly has basically the same underlying reason, but that they can get lost is something I don't think is the case any longer --- would be interesting to see an example. You don't have that museum of footnote confusions any more do you? They would be rather useful as test cases if we ever rework that area.

seasonal greetings

**Comment by car222222** [comment]
> but that they can get lost is something I don't think is the case any longer --- 
> would be interesting to see an example. 

Sure, I think they cannot disappear in the sense of not being typeset, but they can possibly still end up not being visible/printed due to ending up outside the ‘crop box’ of the page. 

> You don't have that museum of footnote confusions any more do you?
> They would be rather useful as test cases if we ever rework that area.

They would indeed but they were only ever on a VAX or DEC I guess, so likely still on a now unreadable back-up tape in some fire-proof vault:-).

I can still visually recall some of the effects, such as getting the second part of a split footnote on the page before the first part!  But reconstructing the input now, no way!

**Comment by wwywong** [comment]
Somewhat related: it appears that back in 2005 Uwe Lück also played with this problem a bit; there's a remnant of it on CTAN. https://ctan.org/pkg/tamefloats  

In particular, Uwe referred to both a recent (as of 2005/04/27) "intense discussion of the matter on the LaTeX-L e-mail list," and also Latex bugs `latex/3208`, [`latex/3370`](https://www.latex-project.org/cgi-bin/ltxbugs2html?category=LaTeX&responsible=anyone&state=open&keyword=&pr=latex%2F3370&search=), `latex/3541`, and [`latex/3719`](https://www.latex-project.org/cgi-bin/ltxbugs2html?category=LaTeX&responsible=anyone&state=open&keyword=&pr=latex%2F3719&search=) in the pre Github bug database.

**Comment by oulenz** [comment]
@FrankMittelbach 

> Part of the solution could be as simple as this:
> 
> ```
> \makeatletter
> \patchcmd\end@float{\@cons\@currlist\@currbox}
>    {\@cons\@currlist\@currbox
>     \global\holdinginserts\@ne}
>     {}{}
>     
> \apptocmd\@specialoutput{\global\holdinginserts\z@}
> 
> \makeatother
> ```

How would this change for a two-column table? I've tried changing `\end@float` -> `\end@dblfloat`, but that doesn't seem to be enough to prevent my footnote from being typeset on the next page.
